airflow/jobs.py

```diff
@@ -1918,119 +1971,131 @@ def _manage_executor_state(self, started):
                     self.logger.error(msg)
                     ti.handle_failure(msg)
 
-    def _execute(self):
+    @provide_session
+    def _get_dag_run(self, run_date, session=None):
         """
-        Runs a dag for a specified date range.
+        Returns a dag run for the given run date, which will be matched to an existing
+        dag run if available or create a new dag run otherwise.
+        :param run_date: the execution date for the dag run
+        :param run_date: datetime
+        :param session:
+        :return:
         """
-        session = settings.Session()
-        DagRun = models.DagRun
-
-        # consider max_active_runs but ignore when running subdags
-        # "parent.child" as a dag_id is by convention a subdag
-        if self.dag.schedule_interval and not self.dag.is_subdag:
-            active_runs = DagRun.find(
-                dag_id=self.dag.dag_id,
+        run_id = BackfillJob.ID_FORMAT_PREFIX.format(run_date.isoformat())
+
+        # check if we are scheduling on top of a already existing dag_run
+        # we could find a "scheduled" run instead of a "backfill"
+        run = DagRun.find(dag_id=self.dag.dag_id,
+                          execution_date=run_date,
+                          session=session)
+
+        if not run:
+            run = self.dag.create_dagrun(
+                run_id=run_id,
+                execution_date=run_date,
+                start_date=datetime.now(),
                 state=State.RUNNING,
                 external_trigger=False,
                 session=session
             )
+        else:
+            run = run[0]
 
-            # return if already reached maximum active runs
```

Comment:
Maybe a followup change is active runs doesn't count backfills? I'm actually not sure what the current logic is